testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
-            error!("Relayer: Failed to stop tenure: {e:?}");
-            return Ok(());
+    /// Get a snapshot for an existing burn chain block given its consensus hash.
+    fn get_block_snapshot_consensus(
+        &self,
+        ch: &ConsensusHash,
+    ) -> Result<BlockSnapshot, NakamotoNodeError> {
+        SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), ch)
+            .map_err(|e| {
+                error!(
+                    "Relayer: failed to get block snapshot for new burn view: {:?}",
+                    e
+                );
+                NakamotoNodeError::SnapshotNotFoundForChainTip
+            })?
+            .ok_or_else(|| {
+                error!("Relayer: failed to get block snapshot for new burn view");
+                NakamotoNodeError::SnapshotNotFoundForChainTip
+            })
+    }
+
+    /// Get the Stacks block ID for the canonical tip.
+    fn get_canonical_stacks_tip(&self) -> StacksBlockId {
+        let (ch, bh) =
+            SortitionDB::get_canonical_stacks_chain_tip_hash(self.sortdb.conn()).unwrap();
+        StacksBlockId::new(&ch, &bh)
+    }
+
+    /// Get the public key hash for the mining key.
+    fn get_mining_key_pkh(&self) -> Option<Hash160> {
+        let Some(ref mining_key) = self.config.miner.mining_key else {
+            return None;
+        };
+        Some(Hash160::from_node_public_key(
+            &StacksPublicKey::from_private(mining_key),
+        ))
+    }
+
+    /// Get the tenure-start block header hash of a given consensus hash.
+    /// For Nakamoto blocks, this is the first block in the tenure identified by the consensus
+    /// hash.
+    /// For epoch2 blocks, this is simply the block whose winning sortition happened in the
+    /// sortition identified by the consensus hash.
+    ///
+    /// `tip_block_id` is the chain tip from which to perform the query.
+    fn get_tenure_bhh(
+        &self,
+        tip_block_id: &StacksBlockId,
+        ch: &ConsensusHash,
+    ) -> Result<BlockHeaderHash, NakamotoNodeError> {
+        let highest_tenure_start_block_header = NakamotoChainState::get_tenure_start_block_header(
+            &mut self.chainstate.index_conn(),
+            tip_block_id,
+            &ch,
+        )
+        .map_err(|e| {
+            error!(
+                "Relayer: Failed to get tenure-start block header for stacks tip {tip_block_id}: {e:?}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?
+        .ok_or_else(|| {
+            error!(
+                "Relayer: Failed to find tenure-start block header for stacks tip {tip_block_id}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?;
+        Ok(BlockHeaderHash(
+            highest_tenure_start_block_header.index_block_hash().0,
+        ))
+    }
+
+    /// Determine the type of tenure change to issue based on whether this
+    /// miner successfully issued a tenure change in the last tenure.
```

Comment:
This docstring is confusing.  The docstring implies that you're looking to see if the given miner (identified by `mining_pkh`) issued a tenure change in the last tenure, but you're not looking at `last_tenure`; you're looking instead at `canonical_tenure`.  Can you explain the difference between these two `BlockSnapshot`s?

Reply:
Yeah, I agree this comment, and this function are a bit confusing. I don't think there was anything new here, just part of some refactoring I did to try to make things less confusing. I'll try to come up with a better comment here.

Reply:
How's this? 2e963e7c80b034207b9cfff7d4aa20cfb30faaf8

Reply:
Better. Thanks!